### DISCUSSION TITLE
Additional checks and missing stuff

### DISCUSSION THREAD

**Discussion by Vollstrecker** [pupnp/pupnp#322] [Ideas]
As I didn't provide a patch today, I want to explain why, and get some opinions what to go for at all and in what order.

Things missing in cmake:
- Doxygen: I don't know how good the outcome is, but this is working in autotools and missing in cmake.
- Symbol visibillity: libtool exports only symbols and functions string with Upnp. cmake atm exports all. I didn't check yet how to mimic this, so my question is how importent this is seen? I know cmake is able to produce a header that will add/remove Extern C on Releases for internal and external usage and takes care of the different plattforms, but this would mean some codechanges and won't be compatible with autotools, so this option can't be considered until we are sure to drop this. There a functions to en-/disable exporting that takes care of the different platform, but I can't test this on Mac, and didn't read enough yet to see if this will help at all.

Thing I planned, but am not sure how helpful they are:
- valgrind: I wanted to add this today, but realised that it won't help, as we have not really bins that use the lib hard enough, maybe with the samples, but (as we lack some docs) I just don't know how to use them.
- clang_tidy, cppcheck and cpplint: All three are easy to include, but we would need to run them on all platforms to make sure we catch the right stuff. And at least clang has many checks that maybe are of no use at all for us.

Opinions?

**Comment by mrjimenez** [comment]
Hi Vollstrecker,

1. Doxygen: it should not be difficult to add a target for CMake. Doxygen is not really part of autotools. The only detail is where you plan to put the generated files. I assume it would be somewhere like build/docs? As always, its up to you, since you understand CMake way better than I do. Doxygen plain vanilla is as easy as this `$ doxygen`. That's it. Run the executable in the `docs` folder and the documentation will be generated to the file doxygen/latex/refman.pdf. We currently rename that file to libupnp-reference.pdf for convenience. The magic lies inside the `docs/Doxyfile` file, where everything can be configured. Maybe your main concern is with the variable `OUTPUT_DIRECTORY`. Well, the sky is the limit for you, you could, e.g., create a .in file and populate the variables you want, creating a new Doxyfile inside the build folder and make the resulting pdf go in there too. But this is just an idea from a person who currently knows very little about the build system.
2. Symbol visibility. If I am not mistaken, this is only a thing for windows dynamic linking. On unixes, EXPORT_SPEC is a macro that expands to nothing. So the visibility is given by the include files that you install in the user's system. I did not understand what CMake has to do with it, since removing the `extern "C"` does not sound like a good thing to do. This is not related to exporting, this is related to the C ABI. Library functions that will be called from either C or C++ must have headers with the `extern "C"` inside a conditional `#ifdef __cplusplus`. So I would advise not to let CMake mess with that. On the other hand, windows DLLs have a particular ABI, so the situation here is different, in fact, you can say it is opposite: whoever wants to call a function inside a DLL must adopt a certain ABI. That has the advantage that you do not care about which language is calling your functions. But has the not so big disadvantage that you have to mark the functions you want to export. On windows maybe there are issues that I am missing, because they have different markers for import and export: you import a function when you want to link with a DLL and you export a function when you are declaring it. It is likely that exporting also creates some glue sections on the object file to take care of the dynamic linking process. Long ago Windows had some legacy dynamic linking modes, like using numbers to access the functions, maybe they have not yet abandoned that. For static linking, EXPORT_SPEC also expands to nothing on WIN32. To summarize, this might be an issue for WIN32 DLLs only. Even then, I think it is unlikely to be a problem since the library headers have been written to take this into account.
3. Valgrind: incredibly usefull. Actually the sanitizer build does that to some extent. The samples work fine, but I understand they need docs that they have never had. On the other hand, very little hacking is needed to understand how they work. Run `./tv_device` in one window and `./tv_ctrlpt` in another. `./tv_device` is the server and sits there. The only command it accepts is `exit`. `./tv_ctrlpt` accepts `help` and you can go from there. It is a fake TV device that is remotely UPNP controlled. The do all the basic `UPNP` functionality.
4. clang_tidy, cppcheck and cpplint: I think they can be useful for devs indeed. If it is easy to add, please do it whenever you have the time.

Regards,
Marcelo.

**Reply by Vollstrecker** [reply]
> Doxygen: it should not be difficult to add a target for CMake.

Nope, the question was more about priority and ordering.

For the visibillity: The extern was just an example, this also handles these DECL_SPEC for import export and creates macros to mark deprecated ones also. I didn't dig deeper into this, just stumpled upon while checking the overall visibility. libtool takes care that only Upnp* which are 599 syms. The cmake one atm exposes also things like sock_make_blocking and overall 1440 syms.

For valgrind and the samples: Here the question was more about how complex will this workflow be, or what has to be thrown at the samples to get a good coverage over the lib? It won't be very useful if this just initialized and destroys the handler.

>If it is easy to add

Inclusion is as easy as the asan-thing. The right value in the right var in front of the configure run, that's all (applies for valgrind also). I checked how it works with include_what_you_use and just thought that additional 3 platform-runs that need to be compared are maybe not that handy as they should be.

Oh, and while talking about 3 platforms. You combined the flags with your learning yaml commits. If you find a way to only add the openssl when on ubuntu, we could combine all 3 into just one job with a big matrix, this would shrink ccpp.yml to half (maybe a bit less). So we would have one workflow everywhere which basically was one of the ideas when cmake was started. I tried that yesterday and failed miserably.

**Reply by mrjimenez** [reply]
Hi,

1. Well, I don't consider a huge priority, since generating the docs can always be done by hand. Also, the pdf is not included in the tarball. So, no hurries here. The command `make docs` is actually longer than `doxygen` :laughing: 
2. Please, show me the command that you use to do this visibility checking. I got very curious now.
3. I don't know how to give you a precise answer. I have used the sanitizer build, which basically adds a valgrind module, several times, usually with the unit tests, but also with the samples, as in the case of the recursive IXML free problem. So the typical workflow is with the unit tests.
4. Then maybe just add where you can test. We can always extend stuff later.
5. I have tried something similar, but I have also failed. The only two places where we can put an if is to test if a job or a step should be executed.  Maybe we could define our variables using the values in the matrix and leave them all defined in the environment. I agree this is slightly ugly, but is the best I could come with.

**Reply by Vollstrecker** [reply]
> The command make docs is actually longer than doxygen

yep, I thought more about generating them just always and only install on request. Don't know how much this would affect the time.

> Please, show me the command that you use to do this visibility checking. I got very curious now.

Just ```readelf -s``` on the .so

>  I don't know how to give you a precise answer....

For the tests it is doable. I just thought about throwing commands at tv_client to activated parts of the lib that are not covered by tests.

> Maybe we could define our variables using the values in the matrix and leave them all defined in the environment

I tried that and wanted to use {{ if: eq...ubuntu.... }} to add open_ssl just there. Failed also. Now I'm thinking about something like cmake/test_options and then just put all defines right there, so we could just run cmake -Dgithub_action=ON or something and do the rest there. Would be 3 lines extra in CMakeLists.txt and a file that just can be ignored for the tarball.

**Reply by Vollstrecker** [reply]
Just tried to bring Mac and Win together. Is upunp_init.log needed every time? Or is there a reason to just create it on Win?

**Reply by mrjimenez** [reply]
Wouldn't it be easier to bring Mac and Unix together? Windows is the different shell.

Do you mean test_init.log? I don't know about upnp_init.log. It is always generated and is needed to see what has happened in UpnpInit2(). These tests should probably be done in a script file, it gets hard to test and make it work using the yaml. The file is only printed when there is an error in the test, so you need shell dependent code to test for the return code of the executable.

**Reply by Vollstrecker** [reply]
> Wouldn't it be easier to bring Mac and Unix together? Windows is the different shell.

Nope, it's cmake <opts>, cmake --build, ctest. The magic happens in there. The problem with unix is the additional open_ssl arg that we only can use there, as win doesn't have it for now, and on mac the headers are missing.

> Do you mean test_init.log?

That was it.

> It is always generated and is needed to see what has happened

Nope, I asked because it didn't exist on Mac. https://github.com/Vollstrecker/pupnp/runs/2326610202?check_suite_focus=true

> These tests should probably be done in a script file

Tomorrow I'll play around a bit with the matrix. Maybe I set the tools from the matrix, or at least a counter that I can react on in the include for the other args. And then I'll check if I can just print the test_init.log as a post-test-command, so we don't have to bother about that in the yml.

**Reply by mrjimenez** [reply]
> Nope, I asked because it didn't exist on Mac. https://github.com/Vollstrecker/pupnp/runs/2326610202?check_suite_focus=true
>
> Users/runner/work/_temp/310543e0-05fe-4981-b513-7cd4d54fe397.sh: line 3: type: upnp/test/test_init.log: not found

The command `type` on a Mac, which is a unix in disguise, is not what you may be thinking. `type` on a windows machine is `cat` on a unix machine. If you use `cat` on a Mac it is likely to work.

The error you see above is not related to the existence of the file. `type` tries to describe the command that follows it, look:

```bash
$ ls
CMakeLists.txt  linux     Makefile.am  tv_combo   tv_device
common          Makefile  Makefile.in  tv_ctrlpt  web

$ type ./CMakeLists.txt
./CMakeLists.txt not found

$ type ls              
ls is an alias for ls --color=tty
```

Just learned about type. We learn new things everyday :laughing:

**Reply by Vollstrecker** [reply]
Everyday something new, let's see what happens tomorrow.

**Reply by mrjimenez** [reply]
LOL.
